versioning/worker-versions.md

```diff
@@ -0,0 +1,174 @@
+  decommissioned
+* Ensure histories replayed on older versions fail fast with informative errors
+* Easy visibility into what versions deployed workers are using
+
+Building on this base set of features, we plan to eventually implement
+automatic safe-rollout of changes, and possibly tools like k8s operators
+which can automatically manage worker deployments.
+
+## Safe-rollout
+
+Intentionally avoiding deep into details here, the foundation laid here should enable
+us, as well as users, to more easily implement automated and safe rollout
+of changes to workers. A process like this one can become the de-facto experience:
+
+* If deploying a known-breaking change, new workflows are only processed by
+  the new version, existing open workflows are processed by old workers.
+* If deploying an expected-to-be-compatible change, new workflow _tasks_ are
+  attempted by new workers, and if they succeed, that workflow will subsequently
+  only be processed by new workers. If they fail, execution goes back to old
+  workers.
+
+# Using opaque whole-worker-identifiers as worker versions
+
+Thanks to @jlegrone for his great feedback encouraging us to revive this approach.
+
+We already have a field called `binaryChecksum` (or similar) in `WorkerOptions`
+today in SDKs. This field's intent is to uniquely identify a specific worker
+build. Since we believe versioning entire workers is the easiest solution, we
+can lean on this existing field.
+
+Deprecating the name `binaryChecksum`and replacing it with something more 
+accurate like `workerBuildId` would be appropriate, since most of our SDKs
+don't even produce a binary.
+
+With this approach, users who wish to make use of the improved versioning 
+features should set the `workerBuildId` during their CI process to a value that
+is unique for that worker's implementation. _Implementation_ rather than 
+_build_ is an important distinction here, since rebuilds of identical
+code should (ideally) produce workers wit the same ID. It is not necessary to
+do this, and it is varyingly difficult depending on the language, but when
+done it enables users to produce workers with the same ID from their CI builds
+if they didn't touch worker code, but something else (ex: The CI definition itself).
+This involves some work on the part of the user, but is more reliable than expecting
+worker authors to increment a version number.
+
+Unfortunately it is infeasible to automatically set an appropriate ID for many languages.
+
+## How is versioning accomplished with the `workerBuildId`?
+
+Every new worker ID will be assumed to be incompatible with the existing id(s)
+unless explicitly specified otherwise. New workers will not process *anything* 
+until explicitly enabled if they opt into build-id based versioning. Users who 
+wish to always have new workers begin processing new workflows can make an API 
+call immediately establishing the new worker as latest as part of their 
+deployment process (see below).
```

Comment:
Trying to envision the easiest user flow.

So in TS, my CI would:

1. Deploy worker code
2. Run a script that:
  - Asks TS SDK what the `workerBuildId` is for the same `WorkerOptions` I just deployed
  - Calls `SetWorkerBuildIdOrdering` without `is_compatible_with` (or SDK can do this as well)

And if I know that it's actually a compatible change, I manually call `SetWorkerBuildIdOrdering` to correct CI's call, with an `is_compatible_with` that I find in `tctl tq describe --tq foo`, which has two new columns, `BUILD_ID` and `FIRST_ACCESS_TIME` (so I know when I deployed it).

How does that sound?

We could avoid step 2 and simplify by deploying code that had an option like:

```ts
Worker.create({
  ...opts,
  useAsLatestVersion: true
})
```

Reply:
Yeah, that's about right

> Asks TS SDK what the workerBuildId is for the same WorkerOptions I just deployed

In principle you may already know what it is, since you're deploying something you built and you set the ID when building, you might just pass along that info to the deploy job - but it's a good point that we want an easy way to "ask" a given worker what it's build ID is.

And yeah, the avoid step 2 part, I think that can pretty easily be done by just having your worker app make that call at the same time it starts up the worker if that's the behavior you want. We can indeed make it a first-class option if people are doing this constantly.

Reply:
> you may already know what it is, since you're deploying something you built and you set the ID when building

Ah okay, I was thinking it was still generated based on code like the checksum.

> And yeah, the avoid step 2 part, I think that can pretty easily be done by just having your worker app make that call at the same time it starts up the worker if that's the behavior you want.

I wouldn't want to just call `SetWorkerBuildIdOrdering` on start, since old deployed workers might get restarted, but I could do:

```
isNewlyDeployed = !GetWorkerBuildIdOrdering()
if isNewlyDeployed
  SetWorkerBuildIdOrdering()
```